File: android/app/src/main/java/com/kryptos/android/core/SmartStego.kt
package com.kryptos.android.core

import java.security.SecureRandom
import java.text.Normalizer

object SmartTextStego {
    private const val MAGIC = 0xC6
    const val MAX_PAYLOAD_BYTES = 0x7FFF

    private val random = SecureRandom()
    private const val RESYNC_STARTS = 3

    private val commaBefore = setOf("but", "so", "yet", "then", "while", "because", "though", "aber", "denn", "sondern", "и", "но", "а", "затем", "потом", "пока", "когда", "поэтому")

    fun encode(data: ByteArray, language: StegoLanguage = StegoLanguage.forSystem()): String =
        encode(data, language, random.nextInt(256))

    internal fun encode(data: ByteArray, language: StegoLanguage, seed: Int): String {
        require(data.size <= MAX_PAYLOAD_BYTES) { "payload too large for the smart stego frame" }
        val g = grammar(language)
        val framed = frameEncode(data, seed)
        val reader = BitReader(framed)
        val bodies = ArrayList<String>()
        while (reader.position < reader.total) {
            val openerIdx = reader.read(g.openerBits)
            val structure = g.structures[g.structOf[openerIdx]]
            val parts = ArrayList<String>(structure.size)
            for (element in structure) {
                when (element) {
                    is Element.Literal -> parts.add(element.word)
                    is Element.Slot -> parts.add(g.slots[element.type][reader.read(g.slotBits[element.type])])
                }
            }
            bodies.add(render(g.openers[openerIdx], g.openerKind[openerIdx], parts))
        }
        return assemble(bodies, seed)
    }

    fun decode(text: String): ByteArray? {
        val tokens = tokenize(text)
        if (tokens.isEmpty()) return null
        for (g in grammars) decode(tokens, g)?.let { return it }
        return null
    }

    fun looksLikeStego(text: String): Boolean = decode(text) != null

    fun mightBeStego(text: String, sample: Int = 6): Boolean {
        val tokens = tokenize(text.take(240))
        if (tokens.size < sample) return false
        val head = tokens.take(8)
        return grammars.any { g -> head.any { g.openerIndex.containsKey(it) } }
    }

    private fun decode(rawTokens: List<String>, g: Grammar): ByteArray? {
        val tokens = rawTokens.filter { it in g.vocab }
        if (tokens.isEmpty()) return null
        var tried = 0
        for (start in tokens.indices) {
            if (!g.openerIndex.containsKey(tokens[start])) continue
            decodeFrom(tokens, start, g)?.let { return it }
            if (++tried >= RESYNC_STARTS) break
        }
        return null
    }

    private fun decodeFrom(tokens: List<String>, start: Int, g: Grammar): ByteArray? {
        var pos = start
        val writer = BitWriter()
        var sentences = 0
        while (pos < tokens.size) {
            val openerIdx = g.openerIndex[tokens[pos]] ?: break
            val pending = ArrayList<IntArray>()
            pending.add(intArrayOf(openerIdx, g.openerBits))
            var next = pos + 1
            var complete = true
            for (element in g.structures[g.structOf[openerIdx]]) {
                when (element) {
                    is Element.Literal ->
                        if (next < tokens.size && tokens[next] == element.word) next++ else complete = false
                    is Element.Slot -> {
                        val idx = if (next < tokens.size) g.slotIndex[element.type][tokens[next]] else null
                        if (idx != null) {
                            pending.add(intArrayOf(idx, g.slotBits[element.type]))
                            next++
                        } else {
                            complete = false
                        }
                    }
                }
                if (!complete) break
            }
            if (!complete) break
            for (item in pending) writer.append(item[0], item[1])
            pos = next
            sentences++
        }
        if (sentences == 0) return null
        return frameDecode(writer.bytes())
    }

    private fun render(opener: String, kind: Int, parts: List<String>): String {
        val sb = StringBuilder()
        sb.append(opener.replaceFirstChar { it.uppercase() })
        if (kind == 0) sb.append(",")
        for (part in parts) {
            if (part in commaBefore) sb.append(",")
            sb.append(" ").append(part)
        }
        return sb.toString()
    }

    private fun assemble(bodies: List<String>, seed: Int): String {
        var x = (seed xor 0x3B) and 0xFF
        fun next(): Int {
            x = (x * 197 + 91) and 0xFF
            return x
        }
        val sb = StringBuilder()
        for ((i, body) in bodies.withIndex()) {
            if (i > 0) sb.append(if (next() % 7 == 0) "\n" else " ")
            sb.append(body)
            sb.append(if (next() % 10 == 9) "!" else ".")
        }
        return sb.toString()
    }

    private fun frameEncode(data: ByteArray, seed: Int): ByteArray {
        val shortLen = data.size <= 0x7F
        val headerSize = if (shortLen) 2 else 3
        val inner = ByteArray(headerSize + data.size + 1)
        inner[0] = MAGIC.toByte()
        if (shortLen) {
            inner[1] = data.size.toByte()
        } else {
            inner[1] = (0x80 or (data.size shr 8)).toByte()
            inner[2] = (data.size and 0xFF).toByte()
        }
        data.copyInto(inner, headerSize)
        inner[inner.size - 1] = crc8(inner, 0, inner.size - 1).toByte()

        val framed = ByteArray(1 + inner.size)
        framed[0] = seed.toByte()
        var x = seed and 0xFF
        for (i in inner.indices) {
            x = (x * 197 + 91) and 0xFF
            framed[1 + i] = (inner[i].toInt() xor x).toByte()
        }
        return framed
    }

    private fun frameDecode(bytes: ByteArray): ByteArray? {
        if (bytes.size < 4) return null
        var x = bytes[0].toInt() and 0xFF
        val inner = ByteArray(bytes.size - 1)
        for (i in inner.indices) {
            x = (x * 197 + 91) and 0xFF
            inner[i] = (bytes[1 + i].toInt() xor x).toByte()
        }
        if ((inner[0].toInt() and 0xFF) != MAGIC) return null
        val b1 = inner[1].toInt() and 0xFF
        val n: Int
        val offset: Int
        if (b1 < 0x80) {
            n = b1
            offset = 2
        } else {
            if (inner.size < 3) return null
            n = ((b1 and 0x7F) shl 8) or (inner[2].toInt() and 0xFF)
            offset = 3
        }
        if (inner.size < offset + n + 1) return null
        if ((inner[offset + n].toInt() and 0xFF) != crc8(inner, 0, offset + n)) return null
        return inner.copyOfRange(offset, offset + n)
    }

    private fun crc8(data: ByteArray, from: Int, to: Int): Int {
        var crc = 0
        for (i in from until to) {
            crc = crc xor (data[i].toInt() and 0xFF)
            repeat(8) {
                crc = if (crc and 0x80 != 0) ((crc shl 1) xor 0x07) and 0xFF else (crc shl 1) and 0xFF
            }
        }
        return crc
    }

    private fun tokenize(text: String): List<String> = StegoTokenizer.split(text)

    private val englishGrammar by lazy { Grammar(SmartStegoData.english) }
    private val russianGrammar by lazy { Grammar(SmartStegoData.russian) }
    private val germanGrammar by lazy { Grammar(SmartStegoData.german) }
    private val grammars by lazy { listOf(englishGrammar, russianGrammar, germanGrammar) }

    private fun grammar(language: StegoLanguage): Grammar =
        when (language) {
            StegoLanguage.RUSSIAN -> russianGrammar
            StegoLanguage.GERMAN -> germanGrammar
            StegoLanguage.ENGLISH -> englishGrammar
        }

    private sealed class Element {
        class Literal(val word: String) : Element()
        class Slot(val type: Int) : Element()
    }

    private class Grammar(raw: SmartStegoData.Grammar) {
        val openers: List<String> = raw.openers.map { nfc(it) }
        val openerKind: List<Int> = raw.openerKind
        val structOf: List<Int> = raw.structOf
        val slots: List<List<String>> = raw.slots.map { list -> list.map { nfc(it) } }
        val openerBits: Int = bits(openers.size)
        val slotBits: List<Int> = slots.map { bits(it.size) }
        val structures: List<List<Element>> = raw.structures.map { row ->
            row.map { token ->
                if (token.startsWith("#")) Element.Slot(token.substring(1).toInt())
                else Element.Literal(nfc(token))
            }
        }
        val openerIndex: Map<String, Int> = openers.withIndex().associate { (i, w) -> w.lowercase() to i }
        val slotIndex: List<Map<String, Int>> = slots.map { list ->
            list.withIndex().associate { (i, w) -> w.lowercase() to i }
        }
        val vocab: Set<String> = buildSet {
            addAll(openers.map { it.lowercase() })
            for (list in slots) addAll(list.map { it.lowercase() })
            for (row in structures) for (element in row) {
                if (element is Element.Literal) add(element.word.lowercase())
            }
        }

        private fun nfc(w: String): String = Normalizer.normalize(w, Normalizer.Form.NFC)

        private fun bits(n: Int): Int {
            require(n > 0 && (n and (n - 1)) == 0) { "grammar list size must be a power of two" }
            return Integer.numberOfTrailingZeros(n)
        }
    }

    private class BitReader(private val bytes: ByteArray) {
        val total: Int = bytes.size * 8
        var position: Int = 0
            private set

        fun read(n: Int): Int {
            var value = 0
            repeat(n) {
                val bit = if (position < total) {
                    (bytes[position ushr 3].toInt() ushr (7 - (position and 7))) and 1
                } else 0
                value = (value shl 1) or bit
                position++
            }
            return value
        }
    }

    private class BitWriter {
        private var out = ByteArray(64)
        private var size = 0
        private var current = 0
        private var count = 0

        fun append(value: Int, n: Int) {
            var i = n - 1
            while (i >= 0) {
                current = (current shl 1) or ((value ushr i) and 1)
                count++
                if (count == 8) {
                    if (size == out.size) {
                        val grown = out.copyOf(out.size * 2)
                        out.fill(0)
                        out = grown
                    }
                    out[size++] = current.toByte()
                    current = 0
                    count = 0
                }
                i--
            }
        }

        fun bytes(): ByteArray {
            if (count > 0) {
                val padded = out.copyOf(size + 1)
                padded[size] = ((current shl (8 - count)) and 0xFF).toByte()
                return padded
            }
            return out.copyOf(size)
        }
    }
}

File: CipherCore/Sources/CipherCore/Steganography/SmartStego.swift
import Foundation

public enum SmartTextStego {
    private static let magic: UInt8 = 0xC6
    public static let maxPayloadBytes = 0x7FFF

    public static func encode(_ data: Data, language: StegoLanguage = .forSystem()) -> String {
        encode(data, language: language, seed: Int.random(in: 0 ... 255))
    }

    static func encode(_ data: Data, language: StegoLanguage, seed: Int) -> String {
        precondition(data.count <= maxPayloadBytes, "payload too large for the smart stego frame")
        let g = grammar(language)
        let framed = frameEncode(data, seed: seed)
        let reader = BitReader(framed)
        var bodies: [String] = []
        while reader.position < reader.total {
            let openerIdx = reader.read(g.openerBits)
            let structure = g.structures[g.structOf[openerIdx]]
            var parts: [String] = []
            parts.reserveCapacity(structure.count)
            for element in structure {
                switch element {
                case .literal(let word):
                    parts.append(word)
                case .slot(let type):
                    parts.append(g.slots[type][reader.read(g.slotBits[type])])
                }
            }
            bodies.append(render(opener: g.openers[openerIdx], kind: g.openerKind[openerIdx], parts: parts))
        }
        return assemble(bodies, seed: seed)
    }

    public static func decode(_ text: String) -> Data? {
        let tokens = tokenize(text)
        guard !tokens.isEmpty else { return nil }
        for g in grammars {
            if let data = decode(tokens: tokens, grammar: g) { return data }
        }
        return nil
    }

    public static func looksLikeStego(_ text: String) -> Bool { decode(text) != nil }

    private static func decode(tokens rawTokens: [String], grammar g: Grammar) -> Data? {
        let tokens = rawTokens.filter { g.vocab.contains($0) }
        guard !tokens.isEmpty else { return nil }
        var pos = 0
        let writer = BitWriter()
        while pos < tokens.count {
            guard let openerIdx = g.openerIndex[tokens[pos]] else { return nil }
            pos += 1
            writer.append(openerIdx, bits: g.openerBits)
            let structure = g.structures[g.structOf[openerIdx]]
            for element in structure {
                switch element {
                case .literal(let word):
                    guard pos < tokens.count, tokens[pos] == word else { return nil }
                    pos += 1
                case .slot(let type):
                    guard pos < tokens.count, let idx = g.slotIndex[type][tokens[pos]] else { return nil }
                    writer.append(idx, bits: g.slotBits[type])
                    pos += 1
                }
            }
        }
        return frameDecode(writer.bytes)
    }

    private static let commaBefore: Set<String> = ["but", "so", "yet", "then", "while", "because", "though", "и", "но", "а", "затем", "потом", "пока", "когда", "поэтому"]

    private static func render(opener: String, kind: Int, parts: [String]) -> String {
        var sentence = opener.prefix(1).uppercased() + opener.dropFirst()
        if kind == 0 { sentence += "," }
        for part in parts {
            if commaBefore.contains(part) { sentence += "," }
            sentence += " " + part
        }
        return sentence
    }

    private static func assemble(_ bodies: [String], seed: Int) -> String {
        var x = (seed ^ 0x3B) & 0xFF
        func next() -> Int {
            x = (x * 197 + 91) & 0xFF
            return x
        }
        var out = ""
        for (i, body) in bodies.enumerated() {
            if i > 0 { out += next() % 7 == 0 ? "\n" : " " }
            out += body
            out += next() % 10 == 9 ? "!" : "."
        }
        return out
    }

    private static func frameEncode(_ data: Data, seed: Int) -> [UInt8] {
        var inner = [UInt8]()
        inner.reserveCapacity(4 + data.count)
        inner.append(magic)
        if data.count <= 0x7F {
            inner.append(UInt8(data.count))
        } else {
            inner.append(UInt8(0x80 | (data.count >> 8)))
            inner.append(UInt8(data.count & 0xFF))
        }
        inner.append(contentsOf: data)
        inner.append(crc8(inner, 0, inner.count))

        var framed = [UInt8]()
        framed.reserveCapacity(1 + inner.count)
        framed.append(UInt8(seed & 0xFF))
        var x = seed & 0xFF
        for byte in inner {
            x = (x * 197 + 91) & 0xFF
            framed.append(byte ^ UInt8(x))
        }
        return framed
    }

    private static func frameDecode(_ bytes: [UInt8]) -> Data? {
        guard bytes.count >= 4 else { return nil }
        var x = Int(bytes[0])
        var inner = [UInt8]()
        inner.reserveCapacity(bytes.count - 1)
        for i in 1 ..< bytes.count {
            x = (x * 197 + 91) & 0xFF
            inner.append(bytes[i] ^ UInt8(x))
        }
        guard inner[0] == magic else { return nil }
        let b1 = Int(inner[1])
        let n: Int
        let offset: Int
        if b1 < 0x80 {
            n = b1
            offset = 2
        } else {
            guard inner.count >= 3 else { return nil }
            n = ((b1 & 0x7F) << 8) | Int(inner[2])
            offset = 3
        }
        guard inner.count >= offset + n + 1 else { return nil }
        guard inner[offset + n] == crc8(inner, 0, offset + n) else { return nil }
        return Data(inner[offset ..< offset + n])
    }

    private static func crc8(_ data: [UInt8], _ from: Int, _ to: Int) -> UInt8 {
        var crc = 0
        for i in from ..< to {
            crc ^= Int(data[i])
            for _ in 0 ..< 8 {
                crc = (crc & 0x80) != 0 ? ((crc << 1) ^ 0x07) & 0xFF : (crc << 1) & 0xFF
            }
        }
        return UInt8(crc)
    }

    private static func tokenize(_ text: String) -> [String] {
        text.precomposedStringWithCanonicalMapping.lowercased()
            .split(whereSeparator: { !$0.isLetter })
            .map(String.init)
    }

    private static let englishGrammar = Grammar(SmartStegoData.english)
    private static let russianGrammar = Grammar(SmartStegoData.russian)
    private static let grammars = [englishGrammar, russianGrammar]

    private static func grammar(_ language: StegoLanguage) -> Grammar {
        language == .russian ? russianGrammar : englishGrammar
    }

    private struct Grammar {
        enum Element {
            case literal(String)
            case slot(Int)
        }

        let openers: [String]
        let openerKind: [Int]
        let structOf: [Int]
        let slots: [[String]]
        let structures: [[Element]]
        let openerBits: Int
        let slotBits: [Int]
        let openerIndex: [String: Int]
        let slotIndex: [[String: Int]]
        let vocab: Set<String>

        init(_ raw: SmartStegoData.Grammar) {
            let normalizedOpeners = raw.openers.map { $0.precomposedStringWithCanonicalMapping }
            let normalizedSlots = raw.slots.map { $0.map { $0.precomposedStringWithCanonicalMapping } }
            openers = normalizedOpeners
            openerKind = raw.openerKind
            structOf = raw.structOf
            slots = normalizedSlots
            openerBits = Grammar.bits(normalizedOpeners.count)
            slotBits = normalizedSlots.map { Grammar.bits($0.count) }
            structures = raw.structures.map { row in
                row.map { token -> Element in
                    if token.hasPrefix("#") { return .slot(Int(token.dropFirst())!) }
                    return .literal(token.precomposedStringWithCanonicalMapping)
                }
            }
            var openerMap = [String: Int](minimumCapacity: normalizedOpeners.count)
            for (i, word) in normalizedOpeners.enumerated() { openerMap[word] = i }
            openerIndex = openerMap
            slotIndex = normalizedSlots.map { list in
                var map = [String: Int](minimumCapacity: list.count)
                for (i, word) in list.enumerated() { map[word] = i }
                return map
            }
            var allWords = Set(normalizedOpeners)
            for list in normalizedSlots { allWords.formUnion(list) }
            for row in structures {
                for element in row {
                    if case .literal(let word) = element { allWords.insert(word) }
                }
            }
            vocab = allWords
        }

        static func bits(_ n: Int) -> Int {
            precondition(n > 0 && (n & (n - 1)) == 0, "grammar list size must be a power of two")
            return n.trailingZeroBitCount
        }
    }

    private final class BitReader {
        private let bytes: [UInt8]
        let total: Int
        private(set) var position = 0

        init(_ bytes: [UInt8]) {
            self.bytes = bytes
            total = bytes.count * 8
        }

        func read(_ n: Int) -> Int {
            var value = 0
            for _ in 0 ..< n {
                let bit = position < total ? Int((bytes[position >> 3] >> (7 - (position & 7))) & 1) : 0
                value = (value << 1) | bit
                position += 1
            }
            return value
        }
    }

    private final class BitWriter {
        private var out = [UInt8]()
        private var current = 0
        private var count = 0

        func append(_ value: Int, bits n: Int) {
            var i = n - 1
            while i >= 0 {
                current = (current << 1) | ((value >> i) & 1)
                count += 1
                if count == 8 {
                    out.append(UInt8(current))
                    current = 0
                    count = 0
                }
                i -= 1
            }
        }

        var bytes: [UInt8] {
            if count > 0 { return out + [UInt8((current << (8 - count)) & 0xFF)] }
            return out
        }
    }
}
